### PR TITLE
feat: add hep planner in sharding sphere optimizer.

### LINKED ISSUE #11834: Add Hep planner in federate optimize
## Add Hep planner in federate optimize

Add Hep planner to do logic optimization in federate-query.
For example, query pushing down, subquery elimination, and so on.

### PR DESCRIPTION
**Pull request by wgy8283335**
Fixes #11834

Changes proposed in this pull request:
- add hep planner in sharding sphere optimizer.

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateContext.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.calcite.config.CalciteConnectionConfig;
+import org.apache.calcite.jdbc.CalcitePrepare;
+import org.apache.calcite.jdbc.CalciteSchema;
+import org.apache.calcite.jdbc.JavaTypeFactoryImpl;
+import org.apache.calcite.tools.RelRunner;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.List;
+
+public class FederateContext implements CalcitePrepare.Context {
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateContext.java:31]: Is it better to define it as a final class? Same question with the next class below.

Comment (tristaZero) [reply shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateContext.java:31]: Do you think `FederateExecuteDataContext` contains something you need in this one?

**Review by tristaZero — CHANGES_REQUESTED** [review]
Hi @guimingyue could you have a look at this PR? Thanks.

**Comment by guimingyue** [comment]
> Hi @guimingyue could you have a look at this PR? Thanks.

Ok, I will review this PR this weekend.

**Comment by tristaZero** [comment]
Hi @wgy8283335 @guimingyue Any progress?

**Comment by guimingyue** [comment]
> Hi @wgy8283335 @guimingyue Any progress?

@tristaZero Sorry for the late reply, I think it is not a good choice for `FederateOptimizer`  to extend from `Prepare` class.

**Comment by wgy8283335** [comment]
> > Hi @wgy8283335 @guimingyue Any progress?
> 
> @tristaZero Sorry for the late reply, I think it is not a good choice for `FederateOptimizer` to extend from `Prepare` class.
@guimingyue I've removed 'Prepare', and use 'Program'.  This would be better.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateContext.java (hunk 1 above)
Comment (wgy8283335) [reply shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateContext.java]: I have modified.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/ShardingSphereOptimizer.java [hunk 2]
```diff
@@ -49,16 +61,27 @@ public RelNode optimize(final String sql) throws SQLParsingException {
...
+        RelNode node = planner.changeTraits(logicPlan, context.getRelConverter().getCluster().traitSet().replace(EnumerableConvention.INSTANCE));
+        RelRoot root = constructRoot(node, resultType);
+        FederateOptimizer optimizer = new FederateOptimizer(planner);
+        RelRoot result = optimizer.optimize(root);
+        return result.rel;
+    }
+
+    private RelRoot constructRoot(final RelNode node, final RelDataType resultType) {
+        final RelDataType rowType = node.getRowType();
+        final List<Pair<Integer, String>> fields = Pair.zip(ImmutableIntList.identity(rowType.getFieldCount()), rowType.getFieldNames());
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/ShardingSphereOptimizer.java:83]: @wgy8283335 Can we remove this useless final modifier?

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/OptimizeContextFactory.java [hunk 3]
```diff
@@ -156,7 +156,7 @@ private void initProperties(final DatabaseType databaseType) {
     }
     
     private RelOptCluster newCluster() {
-        RelOptPlanner planner = new VolcanoPlanner();
+        RelOptPlanner planner = new VolcanoPlanner(null, null);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/OptimizeContextFactory.java:159]: @wgy8283335 Why add null param?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/OptimizeContextFactory.java (hunk 3 above)
Comment (wgy8283335) [reply shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/OptimizeContextFactory.java:159]: At first, I thought they are different.
After I've checked these methods, they are the same. 
I will fix this.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/ShardingSphereOptimizer.java (hunk 2 above)
Comment (wgy8283335) [reply shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/ShardingSphereOptimizer.java:83]: ok.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateOptimizer.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.calcite.adapter.enumerable.EnumerableConvention;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelRoot;
+import org.apache.calcite.tools.Program;
+import org.apache.calcite.tools.Programs;
+
+@AllArgsConstructor
+public class FederateOptimizer {
```
[the remark this answers is not in the document]
Comment (wgy8283335) [reply shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateOptimizer.java:31]: ok.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateOptimizer.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.calcite.adapter.enumerable.EnumerableConvention;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelRoot;
+import org.apache.calcite.tools.Program;
+import org.apache.calcite.tools.Programs;
+
+@AllArgsConstructor
+public final class FederateOptimizer {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateOptimizer.java:31]: @wgy8283335 Can we merge this class into `ShardingSphereOptimizer`?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateOptimizer.java (hunk 5 above)
Comment (wgy8283335) [reply shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/prepare/FederateOptimizer.java]: OK.

**Review by strongduanmu — APPROVED** [review]

**Review by tristaZero — APPROVED** [review]

**Pull request merged**